### PR TITLE
Online catchup no longer waits for trigger ledger

### LINKED ISSUE #3622: config: consider keeping more ledgers in-memory to improve developer experience
@leighmcculloch pointed out a potential improvement to the experience running core: currently, we store 12 ledgers (~1 minute of consensus) in memory by default. If a new node joins the network with its LCL too far behind, it will have to wait until the next checkpoint publish, which can take minutes. However, if we store a checkpoint worth of ledgers in-memory, we shouldn't have this problem, because new nodes would apply previous checkpoint from the archives, then replay ledgers buffered from the network, minimizing delay to get in sync. This issue tracks a few things:
* Verifying that current core implementation works correctly and allows to catchup without waiting (this should be the case already due to the buffered ledgers trimming logic added to catchup not too long ago)
* Analyzing memory profile of core with the updated config. Each in-memory ledger stores quite a bit of data, so we should evaluate memory growth patterns. If it's tolerable, we should consider increasing how many ledgers are kept. 
* I _think_ this change should also improve liveness and recovery guarantees during catastrophic events (for example, if a validator in a v-blocking set goes down and requires new-db)

### PR DESCRIPTION
**Pull request by SirTyson**
# Description

Resolves #3622

This PR allows out-of-sync nodes to begin catchup immediately instead of waiting for a trigger ledger. This change reduces validator startup time by up to 5 minutes. To accomplish this, validators now save SCP messages for the first ledger on the current checkpoint. These messages contain the hash of the last ledger on the most recently completed checkpoint. When these messages are sent to out-of-sync nodes, this hash is used to immediately begin catchup up to the last completed checkpoint. After this "early catchup" finishes, we either apply the buffer ledgers, or run a 2nd smaller catchup if there is a gap between the first buffered ledger and the ledger that the "early catchup" reached.

While this change decreases validator startup time, it does slightly increase bandwidth requirements. SCP state now includes one additional ledger of information that validators must keep in memory and send to each other.

# Checklist
- [x] Reviewed the [contributing](https://github.com/stellar/stellar-core/blob/master/CONTRIBUTING.md#submitting-changes) document
- [x] Rebased on top of master (no merge commits)
- [x] Ran `clang-format` v8.0.0 (via `make format` or the Visual Studio extension)
- [x] Compiles
- [x] Ran all tests
- [ ] If change impacts performance, include supporting evidence per the [performance document](https://github.com/stellar/stellar-core/blob/master/performance-eval/performance-eval.md)

### CODE REVIEW

**Review by marta-lokhova — CHANGES_REQUESTED** [review]
Did an initial pass, and had a couple of questions. I think we can simplify a few things as well. Haven't looked at the tests yet, as the Herder changes are fairly tricky and we need to carefully think about the edge cases. Welcome to (infamous) Herder!

File: src/catchup/CatchupManagerImpl.cpp [hunk 1]
```diff
@@ -186,35 +197,70 @@ CatchupManagerImpl::processLedger(LedgerCloseData const& ledgerData)
                                    "ledger {:d} was closed on network"),
                         lastLedgerInBuffer);
 
-        startOnlineCatchup();
+        startOnlineCatchup(false);
```
[the remark this answers is not in the document]
Comment (SirTyson) [reply src/catchup/CatchupManagerImpl.cpp:225]: The main issue is that we're not sending LedgerCloseData for the checkpoint ledger, but an SCP message. When the out of sync node receives the SCP message, it does not trigger `processLedger` because it does not receive a ledger that needs to be buffered. The special case is required for early catchup so that catchup knows to check the SCP messages for the relevant hash instead of checking the buffered ledgers. This might be bad design on my part, should we be sending LedgerCloseData instead of SCP messages for the checkpoint ledger? I also might be missing something here because I'm still not well versed on how SCP messages become LedgerCloseData to be processed and buffered.

[the remark this answers is not in the document]
Comment (marta-lokhova) [reply src/catchup/CatchupManagerImpl.cpp:225]: Oh so a bit of background on how nodes externalize. Sending SCP messages is the right thing to do here. When enough SCP messages to reach consensus are received, and processed by the SCP state machine, we can officially "externalize" that ledger. Externalize means your quorum reached consensus, but you might not necessarily be in the right state to apply that ledger (e.g. last applied ledger is N, but N+5 was externalized). This information is passed down to Herder and LedgerManager, which tries to either apply that ledger or start catchup to it. See `HerderImpl::processExternalized`, it constructs `LedgerCloseData`, which is then passed to catchup manager to potentially trigger catchup. So we should be able to buffer that ledger just like any other ledger. 

I'd check that the checkpoint ledger we receive is actually externalized by Herder and passed down to CatchupManager. One other thing to note here is that we trim buffered ledgers once in a while in CatchupManager (see `CatchupManagerImpl::trimSyncingLedgers`). I don't think it should affect the desired buffering logic here, but is it possible that the test isn't setup correctly such that the checkpoint ledger is trimmed?

File: src/overlay/test/OverlayTests.cpp [hunk 2]
```diff
@@ -464,7 +464,8 @@ TEST_CASE("outbound queue filtering", "[overlay][connections]")
             constructSCPMsg(env);
             peer->addMsgAndMaybeTrimQueue(constructSCPMsg(env));
         }
-        REQUIRE(scpQueue.empty());
+        // Always keep checkpoint messages
```
Comment (marta-lokhova) [inline src/overlay/test/OverlayTests.cpp:467]: Could we also test that this queue properly purges checkpoints ledgers (so scpQueue is empty at the end)? Should be a tiny change, like setting `env.statement.slotIndex` to something really old.

**Review by marta-lokhova — COMMENTED** [review]
Thanks for the updates, looks close! I think we just to need to make a small update to the checkpoint sending logic.

File: src/herder/HerderImpl.cpp [hunk 3]
```diff
@@ -726,19 +734,44 @@ HerderImpl::sendSCPStateToPeer(uint32 ledgerSeq, Peer::pointer peer)
...
+    // other messages first, then send checkpoint messages after node that is
+    // catching up knows network state
+    if (highestSlot - mApp.getConfig().MAX_SLOTS_TO_REMEMBER - 1 > checkpoint)
+    {
+        delayCheckpoint = true;
+    }
+
+    // Send MAX_SLOTS_TO_SEND slots
     getSCP().processSlotsAscendingFrom(ledgerSeq, [&](uint64 seq) {
+        if (seq == checkpoint && delayCheckpoint)
```
[the remark this answers is not in the document]
Comment (SirTyson) [reply src/herder/HerderImpl.cpp:762]: I think this is left over from when I was doing some weird things with CatchupManager that I missed during the simplification pass, I'll take it out

[the remark this answers is not in the document]
Comment (SirTyson) [reply src/herder/HerderImpl.cpp:762]: Sorry, now I remember why we do this. We special case message filtering if the message is a checkpoint ledger (see `src/herder/HerderImpl.cpp:628`). This checkpoint ledger is based on `trackingConsensusLedgerIndex`. However, if the checkpoint message is the first message sent, the node does not yet have the correct `trackingConsensusLedgerIndex`, so it cannot determine that the message is the special case checkpoint message that should not be filtered. I'll add a better comment explaining this, if it took me a little while to remember this it definetely needs some better documentation. This delay makes it so that the receiving node probably has the correct `trackingConsensusLedgerIndex` by the time it receives the checkpoint message

[the remark this answers is not in the document]
Comment (marta-lokhova) [reply src/herder/HerderImpl.cpp:762]: hmm, why is this a problem though? as long as we send sequential ledgers without gaps, the delay shouldn't matter, right?

[the remark this answers is not in the document]
Comment (SirTyson) [reply src/herder/HerderImpl.cpp:762]: If we send sequential ledgers, the checkpoint message always gets sent first. This gets thrown out though since the close time of the checkpoint ledger is larger than `MAXIMUM_LEDGER_CLOSETIME_DRIFT`. We need at least one message that won't get filtered out by `MAXIMUM_LEDGER_CLOSETIME_DRIFT` to give us a baseline ledger number to calculate the correct checkpoint ledger so we can special case the filtering for messages for that checkpoint ledger

[the remark this answers is not in the document]
Comment (marta-lokhova) [reply src/herder/HerderImpl.cpp:762]: >If we send sequential ledgers, the checkpoint message always gets sent first

Just to clarify, it's not always first, right? It could be in the middle of the sequence of ledgers that we're sending

> This gets thrown out though since the close time difference of the checkpoint ledger is larger than MAXIMUM_LEDGER_CLOSETIME_DRIFT. We need at least one message that won't get filtered out by MAXIMUM_LEDGER_CLOSETIME_DRIFT to give us a baseline ledger number to calculate the correct checkpoint ledger so we can special case the filtering for messages for that checkpoint ledger. The delay gives us reasonable probability that at least one other message is processed before the checkpoint message is received

Ohh I see, it's the close time drift (at first I was only thinking about ledger numbers, where we discard ledgers with a gap). Interesting, so the closetime drift is 50 seconds, yet by default we ask other nodes to send us is the last 12 slots, or ~60 seconds of consensus. So we basically reject a few oldest slots anyway (regardless of checkpoint logic), which doesn't quite make sense. I'm inclined to update `MAXIMUM_LEDGER_CLOSETIME_DRIFT` to be expressed in terms of MAX_SLOTS_TO_REMEMBER and average ledger close time. But we don't have to do it in this PR; even then, there will still be a possibility that checkpoint ledger is rejected. So let's just keep the delay then 👍

File: src/herder/HerderImpl.cpp [hunk 4]
```diff
@@ -726,19 +734,44 @@ HerderImpl::sendSCPStateToPeer(uint32 ledgerSeq, Peer::pointer peer)
...
+        m.type(SCP_MESSAGE);
+        m.envelope() = e;
+        auto mPtr = std::make_shared<StellarMessage const>(m);
+        peer->sendMessage(mPtr, log);
+        return true;
+    };
+
+    bool delayCheckpoint = false;
+    auto checkpoint = getMostRecentCheckpointSeq();
+    auto highestSlot = getSCP().getHighestKnownSlotIndex();
```
[the remark this answers is not in the document]
Comment (SirTyson) [reply src/herder/HerderImpl.cpp:749]: What is the best way to change this then? Here I want to determine if the checkpoint ledger is included in `MAX_SLOTS_TO_REMEMBER` or not. If it's included in `MAX_SLOTS_TO_REMEMBER`, we don't need a delay. I don't see a great way to access that information looking through the SCP class interface

[the remark this answers is not in the document]
Comment (SirTyson) [reply src/herder/HerderImpl.cpp:749]: Is `checkpoint > trackingConsensusLedgerIndex - MAX_SLOTS_TO_REMEMBER` a valid check for this?

File: src/herder/HerderImpl.cpp [hunk 5]
```diff
@@ -746,19 +746,27 @@ HerderImpl::sendSCPStateToPeer(uint32 ledgerSeq, Peer::pointer peer)
 
     bool delayCheckpoint = false;
     auto checkpoint = getMostRecentCheckpointSeq();
-    auto highestSlot = getSCP().getHighestKnownSlotIndex();
+    auto firstSequentialLedgerSeq =
```
Comment (marta-lokhova) [inline src/herder/HerderImpl.cpp:749]: Herder tracking can be less than MAX_SLOTS_TO_REMEMBER, so we need to handle that (by setting firstSequentialLedgerSeq to LedgerManager::GENESIS_LEDGER_SEQ)

**Review by marta-lokhova — APPROVED** [review]
small bugfix needed, otherwise LGTM!

**Review by mbsdf — APPROVED** [review]

File: src/herder/HerderImpl.cpp [hunk 6]
```diff
@@ -726,19 +734,55 @@ HerderImpl::sendSCPStateToPeer(uint32 ledgerSeq, Peer::pointer peer)
     ZoneScoped;
     bool log = true;
     auto maxSlots = Herder::LEDGER_VALIDITY_BRACKET;
+
+    auto sendSlot = [=](SCPEnvelope const& e, bool log) {
```
Comment (marta-lokhova) [inline src/herder/HerderImpl.cpp:738]: I missed this during my earlier review, but I think this can be problematic on shutdown: Peer is kept alive via shared_ptr, so we still call sendMessage. The issue is that half of the application is destroyed at that point, so there are no guarantees that things will fail gracefully. The reason this was not an issue in master is because the lambda was executed synchronously, whereas with this change we're scheduling a callback for later execution (possibly after shutdown is called).  To make this safe:
* Use weak_ptr, no-op if peer is destroyed by the time lambda is called
* No need to capture everything by value, you might accidentally capture other strong references as the code evolves, which is risky when dealing with async tasks. Safer to explicitly only capture the weak_ptr.

**Comment by marta-lokhova** [comment]
Also, we can squash the commits now in preparation for merge

**Comment by marta-lokhova** [comment]
Looks good!

**Review by mbsdf — APPROVED** [review]

**Comment by graydon** [comment]
r+ f17aa0b

**Comment by marta-lokhova** [comment]
r+ 876a374de738101b116e66d89a3ceb7bfa0b7479

**Comment by marta-lokhova** [comment]
there's a failure in "disconnected topology recovery" test, but there doesn't seem to be anything in this PR that would affect it. I also can't reproduce the failure locally (using the same rng-seed). This is a new-ish test, so possible that we need to tune the timeout a bit. Could you try 120 seconds instead of 90?

**Comment by SirTyson** [comment]
120 seconds failed but 180 seems to work. Both 90 and 120 both pass locally.

**Comment by marta-lokhova** [comment]
r+ 9f029fe0c9193abd77adf495d30737be09b960ba
